### ISSUE TITLE
Eager loading nested Temporal models through find_revision builds incorrect join query

### ISSUE THREAD

**Issue by iturgeon** [fuel/orm#228]
Given a has_many relationship between 2 temporal models, if the parent exists and there are no children within the defined timeframe, find_revision with eager loading of the related children will fail to load the parent.

Reproduce:

``` php
<?
class Model_Parent extends Orm\Model_Temporal
{

    protected static $_properties = [
        'id',
        'name',
        'temporal_start',
        'temporal_end',
    ];

    protected static $_has_many = ['kids'];
}

class Model_Kid extends Orm\Model_Temporal
{

    protected static $_properties = [
        'id',
        'parent_id',
        'name',
        'temporal_start',
        'temporal_end',
    ];

    protected static $_belongs_to = ['parent'];
}

$parent = Model_Parent::forge(['name'=>'Brandon']);
$parent->save();

Model_Parent::find_revision(1, 2147483646, ['kids']);
//PHP Fatal error:  Call to a member function set_lazy_timestamp() on a non-object in /mnt/obo/fuel/packages/orm/classes/model/temporal.php on line 158

```

The error results because the query returns no rows.  It returns nothing because the join query places the time restrictions in the where clause instead of the on clause.  This prevents the join from returning null values for kids.

Query built by Temporal_Model:

``` mysql
SELECT `t0`.`id` AS `t0_c0`,
 `t0`.`name` AS `t0_c1`,
 `t0`.`temporal_start` AS `t0_c2`,
 `t0`.`temporal_end` AS `t0_c3`,
 `t1`.`id` AS `t1_c0`,
 `t1`.`name` AS `t1_c1`,
 `t1`.`temporal_start` AS `t1_c2`,
 `t1`.`temporal_end` AS `t1_c3`
FROM (
    SELECT `t0`.`id`,
        `t0`.`name`,
        `t0`.`temporal_start`,
        `t0`.`temporal_end`
    FROM `parents` AS `t0`
    WHERE `t0`.`id` = 1
    AND `t0`.`temporal_start` <= 2147483646
    AND `t0`.`temporal_end` > 2147483646 LIMIT 1
) AS `t0`
LEFT JOIN `kids` AS `t1`
ON (`t0`.`id` = `t1`.`parent_id`)
WHERE `t1`.`temporal_start` <= 2147483646
    AND `t1`.`temporal_end` > 2147483646
```

I believe the fix is to change it to:

``` mysql
SELECT `t0`.`id` AS `t0_c0`,
 `t0`.`name` AS `t0_c1`,
 `t0`.`temporal_start` AS `t0_c2`,
 `t0`.`temporal_end` AS `t0_c3`,
 `t1`.`id` AS `t1_c0`,
 `t1`.`name` AS `t1_c1`,
 `t1`.`temporal_start` AS `t1_c2`,
 `t1`.`temporal_end` AS `t1_c3`
FROM
(
    SELECT `t0`.`id`,
        `t0`.`name`,
        `t0`.`temporal_start`,
        `t0`.`temporal_end`
    FROM `parents` AS `t0`
    WHERE `t0`.`id` = 1 
    AND `t0`.`temporal_start` <= 2147483646 
    AND `t0`.`temporal_end` > 2147483646
    LIMIT 1
) AS `t0`
LEFT JOIN `kids` AS `t1`
ON
(
    `t0`.`id` = `t1`.`parent_id` 
    AND `t1`.`temporal_start` <= '2147483646' 
    AND `t1`.`temporal_end` > '2147483646'
)
```

Proposed fix:

``` php
    // Original function in Temporal_Query
    protected function modify_join_result($join_result, $name)
    {
        if ( ! is_null($this->timestamp))
        {
            //Add the needed conditions to allow for temporal-ness
            $join_result[$name]['where'][] = array($this->timestamp_start_col, '<=', $this->timestamp);
            $join_result[$name]['where'][] = array($this->timestamp_end_col, '>', $this->timestamp);
        }

        return $join_result;
    }

    // Proposed changes to Temporal_Query
    protected function modify_join_result($join_result, $name)
    {
        if ( ! is_null($this->timestamp))
        {
            //Add the needed conditions to allow for temporal-ness
            $table      = $join_result[$name]['table'][1];
            $query_time = \DB::escape($this->timestamp);
            $join_result[$name]['join_on'][] = array("$table.$this->timestamp_start_col", '<=', $query_time);
            $join_result[$name]['join_on'][] = array("$table.$this->timestamp_end_col", '>', $query_time);
        }

        return $join_result;
    }
```

I'll button up a pull request if this looks ok, I suspect this will need attention from @stevewest

**Comment by iturgeon** [comment]
FYI, I'm not 100% sure if I can rely on looking up the table alias using `$join_result[$name]['table'][1]`

**Comment by emlynwest** [comment]
I'll take a proper look at this tomorrow when it's not 22:30 ;) But I notice that you are using the max timestamp, what happens if you -1 off that? I never intended the latest revision to be found using the max timestamp.

Thanks for the name mention too :)

**Comment by iturgeon** [comment]
max is `2147483647`, so its -1 in the example.   Also my pull request generated a new issue for some reason.  Still navigating the complexities of pull requests / issues  (issue #229)

**Comment by emlynwest** [comment]
Just started looking into this and after applying the fix from #230 I am not able to recreate the issue. Using `Model_Parent::find_revision(1, 2147483646, ['kids'])` I get a valid Model_Parent back with an empty array for the `kids` relation as is expected.

**Comment by iturgeon** [comment]
Very strange.  I checked out a brand new copy to isolate it from my current project.  I was able to reproduce this right away.

I'm testing this via the oil console btw, but heres all the sub package commits (1.6/develop head for everything):
- fuel/fuel https://github.com/fuel/fuel/tree/781b1c966750ef9ab637d842e15d2e6f8f15834e
- fuel/core https://github.com/fuel/core/tree/7215635430e703acbbb0c9883b51e0fecdc5fd8c
- fuel/oil https://github.com/fuel/oil/tree/bcfce0106ee5b3df9f47c3bb7969d37d0658533b
- fuel/orm https://github.com/fuel/orm/tree/a0d1cccd22cae05625bb66153857aa380a71bcea

Software Versions:
- Ubuntu 12.0.4
- Mysql 5.5.29
- PHP 5.4.11

Tables:

``` mysql
CREATE TABLE `kids` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `parent_id` int(11) DEFAULT NULL,
  `temporal_start` int(11) NOT NULL DEFAULT '0',
  `temporal_end` int(11) NOT NULL DEFAULT '0',
  PRIMARY KEY (`id`,`temporal_start`,`temporal_end`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

CREATE TABLE `parents` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `temporal_start` int(11) NOT NULL DEFAULT '0',
  `temporal_end` int(11) NOT NULL DEFAULT '0',
  PRIMARY KEY (`id`,`temporal_start`,`temporal_end`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```

**Comment by emlynwest** [comment]
I had a fresh install too, very odd. I'll give it another go tomorrow and poke around some more.

**Comment by iturgeon** [comment]
Had a co-worker run through the hoops of setting this up:
- php 5.3.2
- mysql 5.1.63

He was able to reproduce with the same error.  :confused:

**Comment by emlynwest** [comment]
Ah! I think I know what it was. I had the bad fix applied to Temporal_Query so the filtering was not kicking in.

**Comment by WanWizard** [comment]
Know that Fuel requires PHP 5.3.3+, so expect more issues with 5.3.2...

**Comment by emlynwest** [comment]
@iturgeon The fix you posted over at #230 seems to have resolved this problem. I am still not able to reproduce this using ubuntu 12.10, php 5.4.6 and mysql 5.5.29

**Comment by iturgeon** [comment]
Thanks for testing it out again @stevewest I'll try again with a blank slate and report back here.

**Comment by iturgeon** [comment]
Ok, last try. I put together a fork of fuel to test this and was able to get a different co-worker to try it out with the same error results.  Heres the repo: https://github.com/iturgeon/fuel (branch: "ahhhhhhhh")

The fix in #230 shouldn't affect this this problem because the related child is a temporal object.

**Comment by emlynwest** [comment]
@iturgeon "good" news! I have been able to reproduce the error with your repo, so I'll take a look at your fix. Thanks again for being such a good sport about all these problems.

**Comment by emlynwest** [comment]
Well I am happy to have your fix over at #229 merged in if @WanWizard is happy with it too.

**Comment by iturgeon** [comment]
Right on!  Thanks for sticking with it :+1:

**Comment by emlynwest** [comment]
Was this fixed with #229?

**Comment by emlynwest** [comment]
@iturgeon Seems like this one has been closed with @WanWizard's commit so I'll close this one for now and it can be reopened if the issue persists.
